### ISSUE TITLE
Pure Python mode does not like -> None annotations

### ISSUE THREAD

**Issue by Anaphory** [cython/cython#3553]
I am trying to change my already heavily typed Python code to run in Cython. For that, it felt useful dropping `@cython.cclass` everywhere and see what happens.

I found out that Cython does not understand the `-> None` function return type annotation as intended. Let's take this dumb function.

    def do_something(l):
        if len(l) == 0:
            return
        l.append(1)

This works fine with the decorator.

```
import cython

@cython.ccall
def do_something(l):
    if len(l) == 0:
        return
    l.append(1)
```

It's type annotation is 

    def do_something(l: List[int]) -> None:

or

    def do_something(l: List[int]) -> type(None):

but with both these type annotations, Cython complains

```
Error compiling Cython file:
------------------------------------------------------------
...
from typing import List, NoneType
import cython

@cython.ccall
def do_something(l: List[int]) -> type(None):
                                     ^
------------------------------------------------------------

test.pyx:5:38: Not a type

Error compiling Cython file:
------------------------------------------------------------
...
import cython

@cython.ccall
def do_something(l: List[int]) -> type(None):
    if len(l) == 0:
        return
       ^
------------------------------------------------------------

test.pyx:7:8: Return value required
```

Cython should know to treat None return types special.

**Comment by scoder** [comment]
Right, would be nice to support the `None` case (the `type(None)` case seems excessive for now). PR welcome.

**Comment by Anaphory** [comment]
I am only now diving into Cython head first, having no C experience and only a recent need to speed up some Python. So I have no idea how to start with a PR for this or anything Cython yet. But I will try to stick around and learn the ropes so I can contribute later.

**Comment by da-woods** [comment]
@Anaphory If you just want to make your code work, you can add `@cython.returns(cython.void)` as a decorator. It'd definitely be nice to infer this from `None` in future though.

**Comment by scoder** [comment]
> @cython.returns(cython.void) 

Note that that's not the same thing because it disables exception propagation. The `-> None` annotation should return at least an error indicator and not `void`.

So, it could actually return `None` as an object, or we could change the signature to Cython's internal return code type (which coerces to Python as `None`), as done for builtin functions in `Builtin.py` (marked as `r`).

**Comment by yudonglin** [comment]
what about -`> NoReturn` ?

**Comment by TeamSpen210** [comment]
NoReturn is something entirely different, it's for a function which never returns, only raises an exception or infinitely loops.

**Comment by Ra2-IFV** [comment]
Hi, in pxd `cdef None foobar()` reports `'None' is not a type identifier`
and for `void` in pxd, `-> None` in py, it becomes `signature mismatching` ...
I'm confused there

**Comment by da-woods** [comment]
@Ra2-IFV that's probably with opening a new issue. I think the decision was that a `None` return type didn't imply void (for exception related reasons).

The easiest fix might be to set `# cython: annotation_typing = False` in the .py file, (on the basis that the types are coming from the pxd file) but that isn't very satisfactory.

**Comment by Ra2-IFV** [comment]
And my linter was not happy if I have `-> cython.void` and don't have a explicit return statement
So the ultimate question is: What should I do for type annotation when a cfunc doesn't have to return a value, like in tutorials?

**Comment by Ra2-IFV** [comment]
@da-woods thank you I'll prepare a new issue for my current situation
